### ISSUE TITLE
Use of Js future inside spawn_blocking code

### ISSUE THREAD

**Issue by faxioman** [cunarist/tokio-with-wasm#12]
Is it possible to use JS Futures in the context of spawn_blocking? A use case is, for example, the use of OPFS, which has a synchronous file handle that can only be used in a web worker. However, to obtain the file handle, two calls that return Promises are necessary.

This is a sample code that should be put inside a spawn_blocking closure:

```rust
let window = web_sys::window().unwrap();
let navigator = window.navigator();
let root: FileSystemDirectoryHandle = JsFuture::from(navigator.storage().get_directory()).await?.into();
let new_file: FileSystemFileHandle = JsFuture::from(root.get_file_handle_with_options(
    file.name().as_str(),
    FileSystemGetFileOptions::new().create(true)
)).await?.into();
let write_handle: FileSystemSyncAccessHandle = JsFuture::from(new_file.create_sync_access_handle()).await?.into();
```

**Comment by temeddix** [comment]
Hi @faxioman :)

I haven't checked if that would be possible, but I would recommend doing async jobs only in the main thread, while running blocking heavy jobs in `spawn_blocking`. Is it possible for you to split the IO logic and calculation logic?

**Comment by faxioman** [comment]
Hi @temeddix!
I've tried to keep the acquisition of the FileSystemSyncAccessHandle in the main thread, utilizing promises/futures for this. Ideally, I would then pass the obtained handle to the spawn_blocking part, especially since the synchronous APIs of OPFS can only be used from a web worker.

However, I encounter the following error:

```
`*mut u8` cannot be sent between threads safely [E0277]
Help: within `{closure@src/lib.rs:121:47: 121:54}`, the trait `Send` is not implemented for `*mut u8`, which is required by `{closure@src/lib.rs:121:47: 121:54}: Send`
Note: required because it appears within the type `PhantomData<*mut u8>`
Note: required because it appears within the type `wasm_bindgen::JsValue`
Note: required because it appears within the type `Object`
Note: required because it appears within the type `FileSystemSyncAccessHandle`
Note: required because it's used within this closure Note: required by a bound in `spawn_blocking`
```

Theoretically, the handle should be sharable between the main thread and the web worker, as I have enabled shared wasm memory. I suspect that the issue may be more related to Rust treating the code inside spawn_blocking as if it were a separate thread, expecting the FileSystemSyncAccessHandle to implement Send.
Unfortunately, I don't have strong skills in Rust yet... but do you think there is something on the Rust side that could allow this type of transfer?

I also want to compliment you on the brilliant approach you have implemented to truly simplify the use of web workers in this context. 💪

**Comment by faxioman** [comment]
My example couldn't work anyway because the create_sync_access_handle method needs to be executed in a worker and returns a promise.

We decided to create a new spawn method that isn't provided by Tokio and that only exists when compiling with wasm bindgen. So:

- _spawn_: uses JS futures (I've noticed the signature changed compared to Tokio, with the removal of Send, since wasm bindgen futures are not Send)
- _spawn_blocking_: uses web workers with callable, no futures
- _spawn_web_worker_ [new method]: uses web workers with futures, allowing the management of those particular cases that require futures in a web worker context. This method is identical to spawn_block, but expects Futures instead of Callables.

With this new method we're able to use opfs sync API 👍
